### PR TITLE
Initial implementation for the JVM and JavaScript.

### PR DESCRIPTION
**Pull request by sjrd**
The Scala Native version is not there yet, as Scala Native does not have yet the necessary mechanisms to implement the API.

The only discrepancy between JVM and JS is the behavior of `protected` constructors. The JVM version considers that they are visible, because they are compiled to `public` in the bytecode, while Scala.js explicitly excludes them. It seems that the most reasonable fix would be to allow access to protected constructors in Scala.js' core `Reflect` mechanism.

### CODE REVIEW

**Comment by sjrd** [comment]
Ready for review, mainly @gzm0.

@densh It would be good if you could have a look as well, mostly to tell us whether any of it is unreasonable for Scala Native. But from our earlier discussions, it should be fine.

File: build.sbt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  )
+))
+
+lazy val reflect = crossProject(JSPlatform, JVMPlatform)
+  .in(file("."))
+  .settings(
+    scalacOptions in (Compile, doc) -= "-Xfatal-warnings",
+  )
+  .jvmSettings(
+    compileOrder := CompileOrder.JavaThenScala,
```
Comment (gzm0) [inline build.sbt]: Why this?

File: build.sbt [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    libraryDependencies += "com.novocode" % "junit-interface" % "0.11" % "test",
```
Comment (gzm0) [inline build.sbt]: Hahahaha, fair enough. I didn't realize that this would introduce a circular dependency between this project and uTest :P

File: jvm/src/main/scala/org/portablescala/reflect/Reflect.scala [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+   *  This method is equivalent to calling
+   *  {{{
+   *  Reflect.lookupLoadableModuleClass(fqcn, Reflect.getClass.getClassLoader)
+   *  }}}
+   *
+   *  @param fqcn
+   *    Fully-qualified name of the module class, including its trailing `$`
+   */
+  def lookupLoadableModuleClass(fqcn: String): Option[LoadableModuleClass] =
+    lookupLoadableModuleClass(fqcn, this.getClass.getClassLoader)
```
Comment (gzm0) [inline jvm/src/main/scala/org/portablescala/reflect/Reflect.scala]: I think I have already mentioned this, but this method is evil: It does not properly delegate to the parent class loaders.

If a class calls `lookupLoadableModuleClass("my.dependency")` it is hostile to being loaded by a private class loader (if the `org.portablescala.reflect` package is loaded in a parent). We really should find a better solution for this :-/

Comment (gzm0) [reply jvm/src/main/scala/org/portablescala/reflect/Reflect.scala]: The only thing I can think of right now to solve this is a macro :(

**Review by gzm0 — CHANGES_REQUESTED** [review]

File: build.sbt (hunk 1 above)
Comment (sjrd) [reply build.sbt]: The default is mixed compilation, where scalac is invoked with the .java source files on its "classpath", and so scalac will compile the API of the .java files to use in the .scala files. After, the .java files are sent to javac with the compiled .class files produced by scalac. This allows both Scala to "see" Java and Java to see Scala. It's a bit slow, though.

In our case, we don't need any Java class to see anything from Scala, so forcing Java-then-Scala is more efficient: first javac is invoked for .java files, then scalac is invoked for .scala files with the .class files produced by javac on its classpath.

tl;dr it's not *necessary*, but it improves compile time.

File: jvm/src/main/scala/org/portablescala/reflect/Reflect.scala (hunk 3 above)
Comment (sjrd) [reply jvm/src/main/scala/org/portablescala/reflect/Reflect.scala]: The macro is probably the only way, yes. It could splice in `this.getClass.getClassLoader` at call site. Hoping that blackbox macros can materialize the `this` of their call site ... (probably they can)

**Comment by sjrd** [comment]
Updated with fixes to make the CI pass on 2.10 and 2.11 (I had misconfigured Travis so I wouldn't *actually* test with those versions).

And with an additional commit that uses macros on the JVM to get the caller's class loader.

File: jvm/src/main/scala/org/portablescala/reflect/Reflect.scala [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  }
+
+  private def load(fqcn: String, loader: ClassLoader): Option[Class[_]] = {
+    try {
+      /* initialize = false, so that the constructor of a module class is not
+       * executed right away. It will only be executed when we call
+       * `loadModule`.
+       */
+      val clazz = Class.forName(fqcn, false, loader)
+      if (inheritsAnnotation(clazz)) Some(clazz)
```
Comment (gzm0) [inline jvm/src/main/scala/org/portablescala/reflect/Reflect.scala:192]: Do we also need to check that the class is public? Note the documentation of `Class.forName`:

> Note that this method does not check whether the requested class is accessible to its caller.

**Review by gzm0 — CHANGES_REQUESTED** [review]

File: jvm/src/main/scala/org/portablescala/reflect/Reflect.scala (hunk 4 above)
Comment (sjrd) [reply jvm/src/main/scala/org/portablescala/reflect/Reflect.scala:192]: Apparently not. Both JVM and JS agree that a private class is discoverable. See the additional commit with a test case.

Comment (gzm0) [reply jvm/src/main/scala/org/portablescala/reflect/Reflect.scala:192]: :O is that really what we want it to do?

Comment (sjrd) [reply jvm/src/main/scala/org/portablescala/reflect/Reflect.scala:192]: There's nothing else we *can* do because in the JVM bytecode such a class is public anyways.

Comment (gzm0) [reply jvm/src/main/scala/org/portablescala/reflect/Reflect.scala:192]: I see... So Scala never emits a non-public class?

**Review by gzm0 — APPROVED** [review]

File: jvm/src/main/scala/org/portablescala/reflect/Reflect.scala (hunk 4 above)
Comment (sjrd) [reply jvm/src/main/scala/org/portablescala/reflect/Reflect.scala:192]: Apparently not ...
